File: frontend/demo/component/menubar/menu-bar-custom-item-data.ts
import 'Frontend/demo/init'; // hidden-source-line
import '@vaadin/menu-bar';
import { html, LitElement } from 'lit';
import { customElement, state } from 'lit/decorators.js';
import type { MenuBarItem, MenuBarItemSelectedEvent } from '@vaadin/menu-bar';
import { applyTheme } from 'Frontend/demo/theme';

@customElement('menu-bar-custom-item-data')
export class Example extends LitElement {
  protected override createRenderRoot() {
    const root = super.createRenderRoot();
    applyTheme(root);
    return root;
  }

  // tag::snippet[]
  @state()
  private items: Array<MenuBarItem<{ value?: string }>> = [
    {
      text: 'Copy',
      children: [
        {
          text: 'Copy as plain text',
          value:
            'Menu Bar\n\nMenu Bar is a horizontal button bar with hierarchical drop-down menus.',
        },
        {
          text: 'Copy as HTML',
          value:
            '<h1>Menu Bar</h1><p>Menu Bar is a horizontal button bar with hierarchical drop-down menus.</p>',
        },
        {
          text: 'Copy as Markdown',
          value:
            '# Menu Bar\n\nMenu Bar is a horizontal button bar with hierarchical drop-down menus.',
        },
      ],
    },
  ];

  protected override render() {
    return html`
      <vaadin-menu-bar
        .items="${this.items}"
        @item-selected="${this.itemSelected}"
      ></vaadin-menu-bar>
    `;
  }

  itemSelected(e: MenuBarItemSelectedEvent<MenuBarItem<{ value?: string }>>) {
    const { value } = e.detail.value;
    if (value) {
      navigator.clipboard.writeText(value);
    }
  }
  // end::snippet[]
}

File: src/main/java/com/vaadin/demo/component/menubar/MenuBarCustomItemData.java
package com.vaadin.demo.component.menubar;

import com.vaadin.demo.DemoExporter; // hidden-source-line
import com.vaadin.flow.component.contextmenu.MenuItem;
import com.vaadin.flow.component.contextmenu.SubMenu;
import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.component.menubar.MenuBar;
import com.vaadin.flow.router.Route;

@Route("menu-bar-custom-item-data")
public class MenuBarCustomItemData extends Div {

    public MenuBarCustomItemData() {
        // tag::snippet[]
        MenuBar menuBar = new MenuBar();
        MenuItem copy = menuBar.addItem("Copy");
        SubMenu copySubMenu = copy.getSubMenu();

        copySubMenu.addItem("Copy as plain text", event -> {
            // Provide a custom value by adding a click listener that holds a
            // reference to that value
            String value = "Menu Bar\n\nMenu Bar is a horizontal button bar with hierarchical drop-down menus.";
            copyToClipboard(value);
        });

        copySubMenu.addItem("Copy as HTML", event -> {
            String value = "<h1>Menu Bar</h1><p>Menu Bar is a horizontal button bar with hierarchical drop-down menus.</p>";
            copyToClipboard(value);
        });

        copySubMenu.addItem("Copy as Markdown", event -> {
            String value = "# Menu Bar\n\nMenu Bar is a horizontal button bar with hierarchical drop-down menus.";
            copyToClipboard(value);
        });
        // end::snippet[]

        add(menuBar);
    }

    private void copyToClipboard(String value) {
        getUI().ifPresent(ui -> ui.getPage()
                .executeJs("window.navigator.clipboard.writeText($0);", value));
    }

    public static class Exporter extends DemoExporter<MenuBarCustomItemData> { // hidden-source-line
    } // hidden-source-line
}
